### ISSUE TITLE
invalid ELF header when using flask-bcrypt

### ISSUE THREAD

**Issue by AlanSwenson** [Miserlou/Zappa#1733]
## Context
ran zappa update in which the only thing I changed was adding "cors": true to zappa_Settings.json and now im getting invalid ELF header because of Bcrypt? causing the deployment to fail

## Expected Behavior
normal update

## Actual Behavior
[1545000508959] Instancing..
[1545000512152] bcrypt is required to use Flask-Bcrypt
[1545000512153] /var/task/bcrypt/_bcrypt.abi3.so: invalid ELF header: ImportError

## Steps to Reproduce
1. import bcrypt into a flask project?

## Your Environment

* Zappa version used: 0.47.1
* Operating System and Python version: OSX Mojave, Python 3.6

## Things I've tried
removing the offending "cors": true
undeploying and deploying
adding bcrypt to my pipfile

**Comment by cl1ent** [comment]
another solution that allows using the current bcrypt version is compiling your own packages by adding `"use_precompiled_packages": false` to your `zappa_settings.json`

this compiles the packages on your machine (or ci environment) and does not use the old versions from [Miserlou/lambda-packages](https://github.com/Miserlou/lambda-packages)

make sure to compile the stuff on a suitable python version and with glibc, musl libc didn't work for me...

**Comment by suminb** [comment]
It appears bcrypt 3.1.5 and above does not provide pre-compiled wheels for `manylinux`. I had the same issue, and had to revert bcrypt back to 3.1.4 :-/

See:
- https://pypi.org/project/bcrypt/3.1.4/#files
- https://pypi.org/project/bcrypt/3.1.5/#files
- https://pypi.org/project/bcrypt/3.1.6/#files

**Comment by charlax** [comment]
I had pinned bcrypt to 3.1.4 but this came up again... Not sure why.

**Comment by prattcmp** [comment]
Still having this problem.

**Comment by bendog** [comment]
I've just updated a previously working python3.6 zappa deployment to python3.7 with the latest zappa and i'm now getting this error too... i'm assuming the version of bcrypt packaged has been updated and we will need to pin a different version

**Comment by charlax** [comment]
My solution was to use a dockerized env to deploy from... This way I don't need to pin packages anymore.

**Comment by prattcmp** [comment]
@charlax that seems more like a workaround than a fix, but I may have to do that

**Comment by charlax** [comment]
You're right, but I'm not sure providing prebuilt up-to-date packages is a core feature of Zappa, so it might make sense anyway, depending on your needs.

**Comment by prattcmp** [comment]
I've spent hours trying to properly configure Docker and there are just too many package problems that I cannot fix. Add on top of that, I keep getting:
Warning! AWS Lambda may not be available in this AWS Region!
Warning! AWS API Gateway may not be available in this AWS Region!

If providing prebuilt up-to-date packages is what Zappa needs to actually work, then I'd argue that it should be a priority of Zappa. Because otherwise, it's not worth the configuration effort.

**Comment by prattcmp** [comment]
I've spent hours trying to configure Docker to work properly with Zappa. The current implementation by Zappa is a hack at best, making it unreasonable to use. Really wanted Zappa to just work instead of causing more headaches than its worth 👎

**Comment by bendog** [comment]
`Python 3.7.3` with pipenv
`zappa==0.48.2`

tried
`bcrypt==3.1.3`
`bcrypt==3.1.4`
`bcrypt==3.1.5`
`bcrypt==3.1.6`

**Comment by prattcmp** [comment]
For anyone who's having these problems, I finally got an implementation that I'm happy with using Docker and PyCharm.

Dockerfile:
`FROM lambci/lambda:build-python3.7

WORKDIR /var/task

# Fancy prompt to remind you are in zappashell
RUN echo 'export PS1="\[\e[36m\]zappa>\[\e[m\] "' >> /root/.bashrc

# Additional RUN commands here
# RUN yum clean all && \
#    yum -y install <stuff>

COPY entrypoint.sh /entrypoint.sh
RUN chmod +x /entrypoint.sh

SHELL ["/bin/bash", "-c"]
ENTRYPOINT source /entrypoint.sh
`

entrypoint.sh:
`#!/usr/bin/env bash

# Enter the virtual environment
echo "CREATING VIRTUAL ENVIRONMENT... PLEASE WAIT..."
python -m venv /var/env
source /var/env/bin/activate

# Install required Python packages (via pip)
echo "INSTALLING PIP PACKAGES..."
pip install -r requirements.txt

# Add source to bash
echo 'source /var/env/bin/activate' >> /root/.bashrc

# Start a new shell
$SHELL
`

**Comment by prattcmp** [comment]
Finally got an implementation that I'm satisfied with using Docker and PyCharm. It creates the virtual environment outside of the bind mount because I was having issues whenever the container was killed while the virtual environment was still running. This alleviates those headaches, and all the other headaches I've been running into quite frankly.

Dockerfile:
```Dockerfile
FROM lambci/lambda:build-python3.7

WORKDIR /var/task

# Fancy prompt to remind you are in zappashell
RUN echo 'export PS1="\[\e[36m\]zappa>\[\e[m\] "' >> /root/.bashrc

# Additional RUN commands here
# RUN yum clean all && \
#    yum -y install <stuff>

COPY entrypoint.sh /entrypoint.sh
RUN chmod +x /entrypoint.sh

SHELL ["/bin/bash", "-c"]
ENTRYPOINT source /entrypoint.sh
```

entrypoint.sh:
```Shell
#!/usr/bin/env bash

# Enter the virtual environment
echo "CREATING VIRTUAL ENVIRONMENT... PLEASE WAIT..."
python -m venv /var/env
source /var/env/bin/activate

# Install required Python packages (via pip)
echo "INSTALLING PIP PACKAGES..."
pip install -r requirements.txt

# Add source to bash
echo 'source /var/env/bin/activate' >> /root/.bashrc

# Start a new shell
$SHELL
```

**Comment by fazla043264** [comment]
still have the same error. anyone found any solution?


> Finally got an implementation that I'm satisfied with using Docker and PyCharm. It creates the virtual environment outside of the bind mount because I was having issues whenever the container was killed while the virtual environment was still running. This alleviates those headaches, and all the other headaches I've been running into quite frankly.
> 
> Dockerfile:
> 
> ```dockerfile
> FROM lambci/lambda:build-python3.7
> 
> WORKDIR /var/task
> 
> # Fancy prompt to remind you are in zappashell
> RUN echo 'export PS1="\[\e[36m\]zappa>\[\e[m\] "' >> /root/.bashrc
> 
> # Additional RUN commands here
> # RUN yum clean all && \
> #    yum -y install <stuff>
> 
> COPY entrypoint.sh /entrypoint.sh
> RUN chmod +x /entrypoint.sh
> 
> SHELL ["/bin/bash", "-c"]
> ENTRYPOINT source /entrypoint.sh
> ```
> 
> entrypoint.sh:
> 
> ```shell
> #!/usr/bin/env bash
> 
> # Enter the virtual environment
> echo "CREATING VIRTUAL ENVIRONMENT... PLEASE WAIT..."
> python -m venv /var/env
> source /var/env/bin/activate
> 
> # Install required Python packages (via pip)
> echo "INSTALLING PIP PACKAGES..."
> pip install -r requirements.txt
> 
> # Add source to bash
> echo 'source /var/env/bin/activate' >> /root/.bashrc
> 
> # Start a new shell
> $SHELL
> ```

I have tried this solution but when I run my docker image I get the following error: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'


although I do have the file in the directory.
